controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,1008 @@
+      for( uint16_t i=0; i<len; i++ ) {
+         int ch = rxDat.Get();
+         if( ch < 0 ) return i;
+         *buf++ = ch;
+      }
+      return len;
+   }
+
+   // Write up to len bytes to the buffer and return the
+   // number actually written.
+   uint16_t write( const char *buf, uint16_t len) {
+
+      uint16_t i;
+      for( i=0; i<len; i++ ){
+         if( !txDat.Put( *buf++ ) )
+            break;
+      }
+
+      // Enable the tx interrupt
+      reg->ctrl[0] |= 0x0080;
+
+      return i;
+   }
+
+   // Return the number of bytes currently in the
+   // receive buffer and ready to be read.
+   int RxFull(){
+      return rxDat.FullCt();
+   }
+
+   // Returns the number of free locations in the
+   // transmit buffer.
+   int TxFree(){
+      return txDat.FreeCt();
+   }
+};
+
+static UART rpUART;
+
+// The UART that talks to the rPi uses the following pins:
+//    PB10 - TX
+//    PB11 - RX
+//    PB13 - RTS
+//    PB14 - CTS
+//
+// These pins are connected to UART3
+static void InitUARTs()
+{
+// NOTE - The UART functionality hasn't been tested due to lack of hardware!
+//        Need to do that as soon as the boards are available.
+   EnableClock( UART3_BASE );
+
+   GPIO_PinAltFunc( GPIO_B_BASE, 10, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 11, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 13, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 14, 7 );
+
+   rpUART.Init( UART3_BASE, 115200 );
+
+   EnableInterrupt( INT_VEC_UART3, 3 );
+}
+
+static void UART3_ISR(){
+   rpUART.ISR();
+}
+
+uint16_t HalApi::serialRead(char *buf, uint16_t len) {
+   return rpUART.read( buf, len );
+}
+
+uint16_t HalApi::serialBytesAvailableForRead() {
+   return rpUART.RxFull();
+}
+
+uint16_t HalApi::serialWrite(const char *buf, uint16_t len) {
+   return rpUART.write( buf, len);
+}
+
+uint16_t HalApi::serialBytesAvailableForWrite() {
+   return rpUART.TxFree();
+}
+
+/******************************************************************
+ * Watchdog timer (see chapter 32 of reference manual).
+ *
+ * The watchdog timer will reset the system if it hasn't been
+ * re-initialized within a specific amount of time.  It's used
+ * to catch bugs that would otherwise hang the system.  When
+ * the watchdog is enabled such a bug will reset the system
+ * rather then let it hang indefinitely.
+ *****************************************************************/
+void HalApi::watchdog_init() {
+   Watchdog_Regs *wdog = reinterpret_cast<Watchdog_Regs *>(WATCHDOG_BASE);
+
+   // Enable the watchdog timer by writing the appropriate value to it's key register
+   wdog->key    = 0xCCCC;
+
+   // Enable register access
+   wdog->key    = 0x5555;
+
+   // Set the pre-scaler to 0.  That setting will cause the watchdog
+   // clock to be updated at approximately 8KHz.
+   wdog->prescale = 0;
+
+   // The reload value gives the number of clock cycles before the
+   // watchdog timer times out.  I'll set it to 1000 which gives
+   // us about 250ms before a reset.
+   wdog->reload = 2000;
```

Comment:
2000 doesn't match the comment.

Also can we put a TODO to choose this number more carefully?  I think a smaller number is probably warranted, as if the device hangs for 250ms, that could be dangerous.  But there are lots of considerations, we don't need to figure it out now...